src/libopensc/card-piv.c

```diff
@@ -3657,88 +3657,138 @@ static int get_keylen(unsigned int alg_id, size_t *size)
 
-	keyfilename = (char *)getenv("PIV_EXT_AUTH_KEY");
-
-	if (keyfilename == NULL) {
-		sc_log(card->ctx,
-			"Unable to get PIV_EXT_AUTH_KEY=(null) for general_external_authenticate");
-		r = SC_ERROR_FILE_NOT_FOUND;
-		goto err;
-	}
-
 	r = get_keylen(alg_id, &expected_keylen);
 	if(r) {
 		sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "Invalid cipher selector, none found for:  %02x", alg_id);
 		r = SC_ERROR_INVALID_ARGUMENTS;
 		goto err;
 	}
 
-	f = fopen(keyfilename, "rb");
-	if (!f) {
-		sc_log(card->ctx, " Unable to load key from file\n");
-		r = SC_ERROR_FILE_NOT_FOUND;
-		goto err;
-	}
-
-	if (0 > fseek(f, 0L, SEEK_END))
-		r = SC_ERROR_INTERNAL;
-	fsize = ftell(f);
-	if (0 > (long) fsize)
-		r = SC_ERROR_INTERNAL;
-	if (0 > fseek(f, 0L, SEEK_SET))
-		r = SC_ERROR_INTERNAL;
-	if(r) {
-		sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "Could not read %s\n", keyfilename);
-		goto err;
-	}
-
-	keybuf = malloc(fsize+1); /* if not binary, need null to make it a string */
-	if (!keybuf) {
-		sc_log(card->ctx, " Unable to allocate key memory");
-		r = SC_ERROR_OUT_OF_MEMORY;
-		goto err;
-	}
-	keybuf[fsize] = 0x00;    /* in case it is text need null */
-
-	if ((readlen = fread(keybuf, 1, fsize, f)) != fsize) {
-		sc_log(card->ctx, " Unable to read key\n");
-		r = SC_ERROR_WRONG_LENGTH;
-		goto err;
-	}
-	keybuf[readlen] = '\0';
-
 	tkey = malloc(expected_keylen);
 	if (!tkey) {
 		sc_log(card->ctx, " Unable to allocate key memory");
 		r = SC_ERROR_OUT_OF_MEMORY;
 		goto err;
 	}
 
-	if (fsize == expected_keylen) { /* it must be binary */
-		memcpy(tkey, keybuf, expected_keylen);
+	p = sbuf;
+	*p++ = 0x5C;
+	*p++ = 0x03;
+	*p++ = 0x2F;
+	*p++ = 0x41;
+	*p++ = 0x4B;
+
+	r2 = piv_general_io(card, 0xCB, 0x3F, 0x00, sbuf, p - sbuf, rbuf, sizeof rbuf);
+	if (r2 >= 0) {
+
+		/* Remove the encompassing outer TLV of 0x53 and get the data */
+		body = sc_asn1_find_tag(card->ctx, rbuf, r2, 0x53, &body_len);
+		if (!body || rbuf[0] != 0x53) {
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "Invalid Key Status Data response of NULL\n");
+			r =  SC_ERROR_INVALID_DATA;
+			goto err;
+		}
+
+		if (body_len < 4) {
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "Invalid Key Status Data must be at least 4 bytes\n");
+			r =  SC_ERROR_INVALID_DATA;
+			goto err;
+		}
+
+		if (body[0] != 0x9B) {
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "Invalid Key Status Data unexpected slot\n");
+			r =  SC_ERROR_INVALID_DATA;
+			goto err;
+		}
+
+		if (body[1] != alg_id) {
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "Invalid Key Status Data unexpected algorithm\n");
+			r =  SC_ERROR_INVALID_DATA;
+			goto err;
+		}
+
+		if (((body[2] << 8) | body[3]) != expected_keylen || body_len != expected_keylen + 4) {
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "Invalid Key Status unexpected length\n");
+			r =  SC_ERROR_INVALID_DATA;
+			goto err;
+		}
+
+		memcpy(tkey, body + 4, expected_keylen);
 	} else {
-		/* if the key-length is larger then binary length, we assume hex encoded */
```

Comment:
Logging the value of `r2` with the info that the env fallback is used would be helpful here.